### PR TITLE
Implement use method for card logic

### PR DESCRIPTION
**Pull request by oskarrough**
Idea is to allow more advanced card logic. Cards should be able to call actions themselves. How exactly it needs to work this we'll try to learn by experimenting in this PR. See https://github.com/oskarrough/slaytheweb/issues/109 for discussion.

### CODE REVIEW

File: public/ui/app.js [hunk 1]
```diff
@@ -201,11 +201,11 @@ export default class App extends Component {
 					</details>
 					<details class="DrawPile Overlay" bottomleft>
 						<summary>Dr<u>a</u>w pile ${state.drawPile.length}</summary>
-						<${Cards} cards=${state.drawPile} />
+						<${Cards} state=${state} />
```
Comment (oskarrough) [inline public/ui/app.js]: this should only show the `drawPile`

File: public/ui/app.js [hunk 2]
```diff
@@ -201,11 +201,11 @@ export default class App extends Component {
...
 					</details>
 					<details class="DiscardPile Overlay" bottomright>
 						<summary align-right>Di<u>s</u>card pile ${state.discardPile.length}</summary>
-						<${Cards} cards=${state.discardPile} />
+						<${Cards} state=${state} />
```
Comment (oskarrough) [inline public/ui/app.js]: this should only show the `discardPile`

Comment (oskarrough) [reply public/ui/app.js]: @SamuelRott

File: public/game/cards.js [hunk 3]
```diff
@@ -95,6 +108,40 @@ export const cards = [
...
+	{
+		name: 'Summer of Sam',
+		type: 'Skill',
+		energy: 0,
+		target: 'player',
+		description: 'Gain 1 HP and draw 2 cards',
+		use(state, target) {
+			let newState = actions.addHealth(state, {target, amount: 1})
+			newState = actions.drawCards(newState, 2)
+			return newState
```
[the remark this answers is not in the document]
Comment (oskarrough) [reply public/game/cards.js]: how would that array look like?

i think if it's "just" an array with strings of actions to call we lose the ability to do your fun custom logic inside each card. For example:

```js
onUse: [
  {action: 'addHealth',amount: 1},
  {action: 'drawCards',amount: 2}
],
```

but if we do something like

```js
onUse() {
  return [
    {action: 'addHealth',amount: 1},
    {action: 'drawCards',amount: 2}
  ]
}
```

you could still have a block of code to do extra logic specific to that card easily.

File: public/ui/app.js [hunk 4]
```diff
@@ -176,7 +176,7 @@ export default class App extends Component {
 					</div>
 
 					<div class="Hand">
-						<${Cards} cards=${state.hand} isHand=${true} energy=${state.player.currentEnergy} />
+						<${Cards} cards=${state.hand} isHand=${true} state=${state} />
```
Comment (oskarrough) [inline public/ui/app.js]: It's a bit confusing to have a prop named `state`. Then we have `this.props.state` vs `this.state` which are notthe same thing. Maybe rename to `gameState`?

File: public/ui/app.js (hunk 2 above)
Comment (oskarrough) [reply public/ui/app.js]: oh you fixed this already

File: public/game/cards.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (oskarrough) [reply public/game/cards.js]: another example, how would you create a card that deals 100 damage only if the target has less than 10hp? Here's a try:

```js
onUse(state, target) {
  if (target.currentHealth > 10) return
  let newState = removeHealth(newState, {target, amount: 100})
  return newState
}

[the remark this answers is not in the document]
Comment (oskarrough) [reply public/game/cards.js]: or taking your condition for the Flourish card an example. Just trying things out here!

```js
{
  ...
  description: 'Gain 5 regen. Can only be played if player is under 50% health',
  conditions: [
    {
      action: 'shouldNotHaveMoreThen',
      percentage: 50
    }
  ],
  condition(state) {
    return checkConditions(this.conditions, state)
  },
},
+ the condtions logic
```

vs

```js
{
  ...
  description: 'Gain 5 regen. Can only be played if player is under 50% health',
  condition(state) {
    const percentage = state.player.currentHealth / state.player.maxHealth * 100
    return percentage < 50
  },
},
```

of course if we start writing the same checks 50 times it'd make sense to refactor into reuseable "condition" methods. But there's also a lot power in just doing some quick one-off logic inside each card.

[the remark this answers is not in the document]
Comment (SamuelRott) [reply public/game/cards.js]: One of the goal of having the conditions like that is that you can/could easily list them, it is not needed with 3 but imagine 15 conditions.
I imagine the creation of new card like and plug and play system, you add some effect and condition to a new card without having to right new code. Could also be in favour of creativity, we can test fun combination without having to code them and so on.

_another example, how would you create a card that deals 100 damage only if the target has less than 10hp?_ => I knew we would need to find a solution for those "action on condition". I will think about it :)

File: public/game/cards.js [hunk 5]
```diff
@@ -90,15 +95,108 @@ export const cards = [
...
+			{
+				action: 'addHealth',
+				parameter: {
+					target: 'playerX',
+					amount: 1
+				}
+			},
+			{
+				action: 'drawCards',
+				preConditions: [
```
Comment (SamuelRott) [inline public/game/cards.js:128]: @oskarrough This is how you add a condition to an action.

File: public/game/cards.js [hunk 6]
```diff
@@ -90,15 +95,108 @@ export const cards = [
...
+	return boolean
+}
+
+function runOnUse(actions, state, target) {
+	let newState = state;
+
+	if (actions && state) {
+		actions.forEach(onUse => {
+
+			if (onUse.preConditions && checkConditions(onUse.preConditions, state)) {
```
Comment (SamuelRott) [inline public/game/cards.js:177]: @oskarrough this is how you would check if an actions as a condition.

**Pull request merged**